### PR TITLE
Trial of Champion fixed to work with latest 3.3.5.

### PR DESCRIPTION
**Pull request by jacekowski**
**Changes proposed:**
As previous PR was abandoned i've picked it up and fixed Wilcalf's PR #21427 to merge cleanly, compile and not crash with latest 3.3.5 branch. 
I'm happy to work on it if there are any issues.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** #383 

**Tests performed:** (Does it build, tested in-game, etc.)
It builds, tested in-game (with single player).

**Known issues and TODO list:** (add/remove lines as needed)
- [ ] Ideally it needs more in game testing.

### CODE REVIEW

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 1]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
+    {
+        if (instance->GetBossState(DATA_ARGENT_CHALLENGE) != SPECIAL &&
+            instance->GetBossState(DATA_ARGENT_CHALLENGE) != DONE)
+            BossAI::EnterEvadeMode();
+    }
+
+    void JustReachedHome() final override
+    {
+        if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == SPECIAL)
+            events.ScheduleEvent(EVENT_CHALLENGE_DONE, 4 * IN_MILLISECONDS);
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: events.ScheduleEvent(EVENT_CHALLENGE_DONE, 4ms);

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 2]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
-        void MovementInform(uint32 MovementType, uint32 /*Data*/) override
+        void JustEngagedWith(Unit* who) override
         {
-            if (MovementType != POINT_MOTION_TYPE)
+            if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == SPECIAL)
                 return;
 
-            instance->SetData(BOSS_ARGENT_CHALLENGE_E, DONE);
+            DoCastAOE(SPELL_VENGEANCE);
+            events.ScheduleEvent(EVENT_RADIANCE, urand(7 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: events.ScheduleEvent(EVENT_RADIANCE, 4ms, 15ms);

if im not wrong initialization like that alrdy exist in TC

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 3]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
+            Talk(SAY_AGGRO_E, who);
+            _JustEngagedWith(who);
+        }
 
-            me->DisappearAndDie();
+        void KilledUnit(Unit* who) override
+        {
+            if (who->GetTypeId() == TYPEID_PLAYER)
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: if (who->IsPlayer*())

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 4]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
+
+                    events.Repeat(15 * IN_MILLISECONDS, 17 * IN_MILLISECONDS);
+                    break;
+                case EVENT_SUMMON_MEMORY:
+                    // Memory spawns at random player's position
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 50.0f, true))
+                        DoCast(target, SPELL_SUMMON_MEMORY, true);
+                    break;
+                case EVENT_MEMORY_AGGRESSIVE:
+                    if(_memory) {
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: move {

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 5]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
+    {
+        return GetTrialOfChampionAI<boss_paletressAI>(creature);
+    }
+};
+
+class npc_argent_soldier : public CreatureScript
+{
+public:
+    npc_argent_soldier() : CreatureScript("npc_argent_soldier") { }
+
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: use RegisterAI ?

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 6]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
+                    me->GetMotionMaster()->MovePoint(POINT_PREFIGHT, homePos);
+            }
+        }
+
+        void JustEngagedWith(Unit* /*who*/) override
+        {
+            switch (me->GetEntry())
+            {
+            case NPC_ARGENT_LIGHWIELDER:
+                _events.ScheduleEvent(EVENT_BLAZING_LIGHT, 10 * IN_MILLISECONDS);
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: 10 * IN_MILLISECONDS -> 10ms  etc.... on all other places

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 7]
```diff
@@ -104,321 +150,710 @@ enum Spells
...
-            MemoryGUID = summon->GetGUID();
+            me->SetReactState(REACT_PASSIVE);
         }
+
+        void JustDied(Unit* /*killer*/) override
+        {
+            me->DespawnOrUnsummon();
+        }
+
+        void AttackStart(Unit* /*who*/) override { }
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: no reasons in empty override functions

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 8]
```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
...
-        void UpdateAI(uint32 uiDiff) override
+    class spell_eadric_hammer_of_righteous_faceroller_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_eadric_hammer_of_righteous_faceroller_SpellScript);
+
+        bool Load() override
         {
-            EscortAI::UpdateAI(uiDiff);
+            Unit* caster = GetCaster();
+            return caster->GetTypeId() == TYPEID_PLAYER && caster->GetMap()->IsHeroic();
```
Comment (soulfrost) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: GetTypeId() == TYPEID_PLAYER  ->  IsPlayer()

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*
- * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: Please cancel this change (restore the deleted line and remove the added one).

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
- * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp]: Please cancel this change (restore the deleted line and remove the added one).

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp [hunk 11]
```diff
@@ -1,5 +1,6 @@
 /*
- * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2006-2009 ScriptDev2 <https://scriptdev2.svn.sourceforge.net/>
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp]: Please cancel this change (restore the deleted line and remove the added ones).

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/instance_trial_of_the_champion.cpp [hunk 12]
```diff
@@ -1,5 +1,5 @@
 /*
- * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/instance_trial_of_the_champion.cpp]: Please cancel this change (restore the deleted line and remove the added one).

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.cpp [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
- * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.cpp]: Please cancel this change (restore the deleted line and remove the added one).

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.h [hunk 14]
```diff
@@ -1,5 +1,6 @@
 /*
- * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ * Copyright (C) 2006-2009 ScriptDev2 <https://scriptdev2.svn.sourceforge.net/>
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.h]: Please cancel this change (restore the deleted line and remove the added ones).

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.h [hunk 15]
```diff
@@ -15,124 +16,284 @@
  * with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
-#ifndef DEF_TOC_H
-#define DEF_TOC_H
+#ifndef TRIAL_OF_CHAMPION_H_
+#define TRIAL_OF_CHAMPION_H_
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.h]: This change looks outdated, is it intentional?

Comment (illfated) [reply src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/trial_of_the_champion.h]: Never mind, I was thinking of something else entirely. (I will delete these 2 comments.)

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp [hunk 16]
```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
...
-            uiWaypoint = 0;
+            return ValidateSpellInfo({ SPELL_HAMMER_RIGHT_DUMMY, SPELL_HAMMER_JUSTICE_STUN });
         }
 
-        InstanceScript* instance;
-
-        uint8 uiWaypoint;
-
-        void WaypointReached(uint32 waypointId, uint32 /*pathId*/) override
+        void HandleDamage(SpellMissInfo missInfo)
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: unused argument
```suggestion
        void HandleDamage(SpellMissInfo /*missInfo*/)
```

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp [hunk 17]
```diff
@@ -746,286 +1471,305 @@ class boss_hunter_toc5 : public CreatureScript
...
-                EnterEvadeMode();
-                bHome = true;
+            void RemoveInvalidTargets(std::list<WorldObject*>& targets)
+            {
+                // The aura should not be applied if there is already a trample aura on target
+                targets.remove_if(Trinity::UnitAuraCheck(true, GetSpellInfo()->Id));
             }
 
-            if (uiPhaseTimer <= uiDiff)
+            void HandleStun(SpellMissInfo missInfo)
```
Comment (illfated) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp]: unused parameter
```suggestion
            void HandleStun(SpellMissInfo /*missInfo*/)
```

**Comment by jackpoz** [comment]
Please check https://circleci.com/gh/TrinityCore/TrinityCore/5698?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link for generated warnings, i.e.:
>[ 68%] Building CXX object src/server/scripts/CMakeFiles/scripts_northrend.dir/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp.o
/home/circleci/project/src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp:957:41: fatal error: unused parameter 'missInfo' [-Wunused-parameter]
        void HandleDamage(SpellMissInfo missInfo)

If you don't use missInfo parameter, you can comment it out like this
```cpp
void HandleDamage(SpellMissInfo /*missInfo*/)
```

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp [hunk 18]
```diff
@@ -15,577 +16,1264 @@
...
+                    }
+                }
+                break;
             }
         }
+    }
+
+    bool CanUseNormalAI()
+    {
+        // must check for unattackable flag to prevent crash in certain situations
```
Comment (jackpoz) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp]: this was commented at https://github.com/TrinityCore/TrinityCore/pull/21427#discussion_r169184672 but it wasn't resolved

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp [hunk 19]
```diff
@@ -15,577 +16,1264 @@
...
-  {746.84f, 623.15f, 411.41f},
-  {747.96f, 620.29f, 411.09f},
-  {750.23f, 618.35f, 411.09f}
-};
-*/
-void AggroAllPlayers(Creature* temp)
+void EnterAggressiveMode(Creature* temp)
 {
-    Map::PlayerList const& PlList = temp->GetMap()->GetPlayers();
+    temp->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_IMMUNE_TO_PC);
```
Comment (jackpoz) [inline src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp]: this was commented at https://github.com/TrinityCore/TrinityCore/pull/21427#discussion_r169184363 but it wasn't resolved

**Comment by jackpoz** [comment]
Could you please go through the feedback provided at https://github.com/TrinityCore/TrinityCore/pull/21427 and implement what was discussed there ?

File: src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp (hunk 2 above)
Comment (ForesterDev) [reply src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp]: Not ms. Seconds

**Comment by MegaserverRU** [comment]
No longer working for latest version 2022/01/15
After fixing the build error

**Pull request closed without merging**